### ISSUE TITLE
veteran.participant_id is not unique (same veteran with 2 file numbers)

### ISSUE THREAD

**Issue by sansb** [department-of-veterans-affairs/caseflow#5429]
SyncIntakeJob is failing because of an issue with participant_id not being unique.

Relevant Sentry error: https://sentry.ds.va.gov/department-of-veterans-affairs/caseflow/issues/1311/

Here's what we are seeing:
- Since https://github.com/department-of-veterans-affairs/caseflow/commit/c7c78c43e077647ecc0ec5abc86f7d72acb0d7d2, Veterans are stored in the Caseflow DB and are fetched from BGS with a `find_or_create_by_file_number` method that first looks for the Veteran in the Caseflow DB and falls back to fetching it from BGS and then creating the record in the Caseflow DB.
- The Caseflow Veteran table has uniqueness constraints on both `file_number` and `participant_id`.
- There exists in BGS a veteran who is represented by two separate `file_number`s, but other things like name and most importantly `participant_id` are the same.
- For this veteran, we have already stored a copy of their BGS record via one of the two `file_number`s. 
- When our SyncIntakeJob tries to sync all of our intakes with external services (BGS and VBMS), it comes across an Intake for this Veteran with the OTHER `file_number`. It attempts to find the veteran by this `file_number` but fails because it does not exist. It then tries to create this veteran, but fails because this veteran's `participant_id` is already in our DB, and that column has a uniqueness constraint. It then returns nil, and we can not access the Veteran record.

Options for fixing this:
- Drop the uniqueness constraint on `participant_id`. This means that our Veteran table could hold multiple pointers to the same veteran. Unclear what the consequences of this would be.
- Change `find_or_create_by_file_number` to add a third attempt to retrieve a veteran, which would be to try getting the veteran's `participant_id` from BGS and finding the Veteran by participant_id.

Open questions:
- What does it mean for a single veteran to have multiple `file_number`s?

**Comment by shanear** [comment]
I like:

> Drop the uniqueness constraint on participant_id. This means that our Veteran table could hold multiple pointers to the same veteran. Unclear what the consequences of this would be.

It's gets to the root of the faulty assumption we originally made and would keep all the same functionality.

**Comment by lowellrex** [comment]
I don't love that our `veterans` table can have two rows with different file numbers but the same BGS participant ID. It seems to me like this table is now modeling _files_ and not _Veterans_ (since we believe the BGS participant ID to more precisely identify the individual Veteran as a person). I think this reduces the general utility and flexibility of this table to contain more information about a Veteran in the future.

Since this change has already been merged, I don't expect that we will change the status quo, but I think either of the two alternatives would allow for greater flexibility of this table going forward.
1. Rename the `veterans` table to something like `files` to increase the precision of the name of the data we are representing.
1. Add a relational table to map multiple file numbers to a single veteran entity and restore the uniqueness constraint of participant ID on the veterans table. Maybe something like the following:
```
CREATE TABLE veterans (
        id serial,
        participant_id varchar(40)
);
CREATE TABLE file_number_for_veteran (
        id serial,
        veteran_id int REFERENCES veterans(id),
        file_number varchar(40)
);
```

**Comment by sansb** [comment]
@lowellrex I think you're totally right- this table is no longer really a Veterans table so much as it is a VeteranFiles table. We opted for the simplest fix since this issue was causing a nightly job to fail and we wanted to fix ASAP. I agree with switching to Veteran/VeteranFile or something like that and think we should do it, but I don't how we will prioritize it. @shanear what do you think?

**Comment by sansb** [comment]
Very interesting. Sounds good.

**Comment by shanear** [comment]
Used this query to find the double file number veterans:
```
Veteran.joins("INNER JOIN veterans AS v ON v.participant_id = veterans.participant_id AND v.id > veterans.id")
```
There are around 50 of them

**Comment by shanear** [comment]
Sent the 58 occurrences of this issue we've found so far to John Dell. Waiting on a response on what to do

**Comment by shanear** [comment]
This is pretty low priority in the list of stuff we are asking BGS for, so I'm gonna push this back to the backlog for a little bit.

**Comment by leahbannon** [comment]
Shane to put Lowell in touch with John to follow up as necessary.

**Comment by shanear** [comment]
@lowellrex The fact that there are two vets to one file number is a bug, confirmed by BGS. For that reason, I'm not sure it's a good idea to build a long term work around to bugs in our system (especially if it involves changing database structure).

We should isolate the work around logic in our app in parallel with putting pressure to fix the underlying bug. That way we can remove the workaround code easily when the underlying bug is fixed. I can put you in touch with John Dell to follow up. I haven't done it because there are many higher priority issues I'm working with john ATM.

**Comment by shanear** [comment]
@mc-usds is submitting a prod ticket for this. I guess we'll wait on a response from that.

**Comment by mc-usds** [comment]
Submitted the ticket: SR-001488

**Comment by mc-usds** [comment]
We learned from Jack that there exists tables that track when:

filenumber or ssn changes: PERSON_IDFCTN_HIST
participant id records that are merged: MERGED_PTCPNT
??didn't catch what this table tracked??: NON_MERGED_PTCPNT

Next step is to get a better understanding how these tables can be used. Both VBMS and Caseflow are interested in accessing these tables to known when a change has been made. VBMS is thinking of address efolders that need to be merged in two parts. First, use the table to address existing efolders associated with a filenumber that has been deleted merged to the correct filenumber. Second part is to develop a solution for handling these scenarios in the future.

**Comment by leikkisa** [comment]
I wanted to add a note that in some cases, the file_number or SSN for a veteran can be used successfully for the some calls to BGS (but not all BGS calls).  The way our code is written, we try the file_number first, and if that doesn't work, we fetch the file_number by SSN.  Due to this, we may have the SSN stored as the file number for some veterans, when there is also a different file number.

**Comment by laurjpeterson** [comment]
@pkarman @leikkisa here's another SSN vs. file number related ticket

**Comment by laurjpeterson** [comment]
This ticket seems related - https://github.com/department-of-veterans-affairs/caseflow/issues/7734

**Comment by pkarman** [comment]
I think we've reached the point where things Just Work, even if data consistency is not optimal. We more or less mirror the inconsistencies that exist upstream.
